src/goto-programs/remove_exceptions.cpp

```diff
@@ -0,0 +1,598 @@
+      type_exprt expr(type);
+      // find the symbol corresponding to the caught exceptions
+      exprt exc_symbol=instr_it->code;
+      while(exc_symbol.id()!=ID_symbol)
+        exc_symbol=exc_symbol.op0();
+
+      binary_predicate_exprt check(exc_symbol, ID_java_instanceof, expr);
+      t_exc->guard=check;
+    }
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_exceptionst::add_function_call_gotos
+
+Inputs:
+
+Outputs:
+
+Purpose: instruments each function call that may escape exceptions
+          with conditional GOTOS to the corresponding exception handlers
+
+\*******************************************************************/
+
+void remove_exceptionst::add_function_call_gotos(
+  const goto_functionst::function_mapt::iterator &func_it,
+  const goto_programt::instructionst::iterator &instr_it,
+  const stack_catcht &stack_catch)
+{
+  assert(instr_it->type==FUNCTION_CALL);
+
+  goto_programt &goto_program=func_it->second.body;
+
+  // save the address of the next instruction
+  goto_programt::instructionst::iterator next_it=instr_it;
+  next_it++;
+
+  code_function_callt &function_call=to_code_function_call(instr_it->code);
+  assert(function_call.function().id()==ID_symbol);
+  const irep_idt &callee_id=
+    to_symbol_expr(function_call.function()).get_identifier();
+
+  if(symbol_table.has_symbol(id2string(callee_id)+EXC_SUFFIX))
+  {
+    // dynamic dispatch of the escaped exception
+    const symbolt &callee_exc_symbol=
+      symbol_table.lookup(id2string(callee_id)+EXC_SUFFIX);
+
+    symbol_exprt callee_exc;
+    callee_exc.set_identifier(id2string(callee_id)+EXC_SUFFIX);
+    callee_exc.type()=callee_exc_symbol.type;
+
+    for(std::size_t i=stack_catch.size(); i-->0;)
+    {
+      for(std::size_t j=stack_catch[i].size(); j-->0;)
+      {
+        goto_programt::targett new_state_pc;
+        new_state_pc=stack_catch[i][j].second;
+        goto_programt::targett t_exc=goto_program.insert_after(instr_it);
+        t_exc->make_goto(new_state_pc);
+        t_exc->source_location=instr_it->source_location;
+        t_exc->function=instr_it->function;
+        // use instanceof to check that this is the correct handler
+        symbol_typet type(stack_catch[i][j].first);
+        type_exprt expr(type);
+        binary_predicate_exprt check_instanceof(
+          callee_exc,
+          ID_java_instanceof,
+          expr);
+        t_exc->guard=check_instanceof;
+      }
+    }
+
+    // add a null check (so that instanceof can ba applied)
+    equal_exprt eq_null(typecast_exprt(callee_exc, pointer_typet()),
+                        null_pointer_exprt(pointer_typet()));
+    // jump to the next instruction
+    goto_programt::targett t_null=goto_program.insert_after(instr_it);
+    t_null->make_goto(next_it);
+    t_null->source_location=instr_it->source_location;
+    t_null->function=instr_it->function;
+    t_null->guard=eq_null;
```

Comment:
Now, after the conditional GOTOs to the exception handlers, there is an unconditional GOTO to the end of the function.